File: project-euler/284/euler_284_v1.py
#!/usr/bin/env python

# The Expat License
#
# Copyright (c) 2017, Shlomi Fish
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import sys
from six import print_
from gmpy2 import mpz

if sys.version_info > (3,):
    long = int
    xrange = range

BASE = 14
powers = [mpz(1)]
for n in xrange(1, 10001):
    powers.append(powers[-1] * BASE)

MP = [[p * d for d in xrange(0, BASE)] for p in powers]

MAX = 9


def rec(n, sq, is_z, digits_sum):
    # print_('sq =', sq)
    s = [[sq, is_z, digits_sum, 0]]
    ret = 0
    while len(s):
        # if len(s) & 0xFF == 0:
        #     print_('n =', len(s))
        if len(s) > MAX+1:
            s.pop()
            continue
        sq, is_z, digits_sum, d = s[-1]
        if (((sq*sq) % powers[len(s)-1]) != sq):
            s.pop()
            continue
        if sq == 0 and len(s) > 3:
            s.pop()
            continue
        if d == 0:
            if not is_z:
                ret += digits_sum
        elif d == BASE:
            s.pop()
            continue
        s[-1][-1] += 1
        s.append([sq+MP[len(s)-1][d], (d == 0), digits_sum+d, 0])
    return ret


# print_(rec(0, long(0), True, 0))
MAX = 10000
print_(rec(0, mpz(0), True, 0))

File: project-euler/284/euler_284_v1.cpp
// The Expat License
//
// Copyright (c) 2017, Shlomi Fish
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <gmpxx.h>
#include <iostream>

const size_t BASE = 14;
const size_t MAX = 10000;

mpz_class MP[10030][BASE];

mpz_class ret = 0;

static void rec(const size_t n, const mpz_class sq, const bool is_z, const size_t digits_sum)
{
    if (n > MAX)
    {
        return;
    }
    if (((sq*sq) % MP[n][1]) != sq)
    {
        return;
    }
    if (sq == 0 and n > 3)
    {
        return;
    }
    if (! is_z)
    {
        ret += digits_sum;
    }
    const auto m = MP[n];
    for (size_t d=0;d<BASE;++d)
    {
        rec(n+1, sq+m[d], (d == 0), digits_sum+d);
    }
}

int main()
{
    for (size_t d=0;d<BASE;++d)
    {
        MP[0][d] = d;
    }
    for (size_t n=1;n<10030;++n)
    {
        for (size_t d=0;d<BASE;++d)
        {
            MP[n][d] = d * MP[n-1][1] * BASE;
        }
    }
    rec(0, 0, true, 0);
    std::cout << ret << std::endl;
}
